asciidoc/volume1/use-cases/tf1-ch-c-use-case-stad.adoc

```diff
@@ -214,6 +225,163 @@ NOTE: This requirement supplements RR1162 in <<ref_ieee_11073_10700_2022>>: _The
+****
+The <<vol1_spec_sdpi_p_actor_somds_provider>> may indicate a timestamp belongs to a specific epoch using the SDPi epoch extension. 
+
+.Notes
+[NOTE]
+[%collapsible]
+====
+Binding timestamps in the <<acronym_mdib>> to a specific epoch may be useful for states that are not updated frequently. 
+
+====
+****
+
+.R1562
+[sdpi_requirement#r1562,sdpi_req_level=shall]
+****
+The <<vol1_spec_sdpi_p_actor_somds_consumer>> shall consider the risks arising from relying on timestamps obtained from different epochs.  
+
+.Notes
+[NOTE]
+[%collapsible]
+====
+It may not be possible to reliably determine the relationship between timestamps obtained from different time-reference frames without addition information regarding the cause of the non-slewing adjustment. For example, if a non-slewing adjustment arises because the device clock was running faster (or slower) than the reference clock then the arithmetic difference between two events spanning the adjustment (even when combined with the step adjustment duration) may not match the elapsed time experienced by an unbiased observer.  
+
+====
+****
+
+
+// This is for the sledge hammer approach. I can't figure out what a universal rule could be or how to communicate epoch changes
+// across MdibVersionGroup/@SequenceId since it seems that any information inside the MDS implicitly is scoped to the 
+// sequence id. 
+.R1566
+[sdpi_requirement#r1566,sdpi_req_level=shall]
+****
+The <<term_manufacturer>> of a <<vol1_spec_sdpi_p_actor_somds_provider>> that changes the MDIB sequence identifier when it can no longer make smooth adjustments to its time-reference frame shall consider the risks arising from gaps in continuous data. 
+
+.Notes
+[NOTE]
+[%collapsible]
+====
+Non-slewing time-adjustments may indicate a serious error that impacts data that has already been:
+ 
+ * displayed on a chart to the user,
+ * exported to other systems.
+
+====
+****
+
+// This may be unneccessary since it applies to all participants from 10700:§5.2.2,RR1162. It does make it clear
+// that epoch versions aren't required though. 
+.R1568
+[sdpi_requirement#r1568,sdpi_req_level=shall]
+****
+The <<term_manufacturer>> of a <<vol1_spec_sdpi_p_actor_somds_provider>> that chooses to omit epoch versions from any timestamp shall consider the risks arising from erroneous timestamps. 
+
+[NOTE]
+[%collapsible]
+====
+Epoch versions may not be required for timestamps on items that update frequently. 
+
+====
+****
+
+// This may be unnecessary as the device could fault at any time. However, perhaps it is useful as a way
+// to surface behaviours as part of conformity statements. And it emphasises the myriad of problems with
+// time steps. 
+.R1569
+[sdpi_requirement#r1569,sdpi_req_level=may]
+****
+A <<vol1_spec_sdpi_p_actor_somds_participant>> may enter a fault state by, for example, setting the `MdsState/@ActivationState` to `Fail` upon detecting a non-slewing time adjustment that it otherwise cannot recover from. 
```

Comment:
Sorry, I'm not sure I completely follow this one and its interaction with other points. I think it depends on whether we put `mustUnderstand="true"` on some extension in the clock descriptor, and whether we are trying to deal with participants outside SDPi. This is where I got to:

If we put `mustUnderstand="true"` on in the clock's descriptor for an versioning extension:
* consumers have to understand whatever we come up with in SDPi otherwise they must not to process messages.
* the 3rd option in 1522 provides more nuance on which timestamps are reliable and which are not:
  * `sdpi:Epochs/@Version` changing signals a abrupt time adjustment,
  * `pm:ClockState/@LastSet` signals when that occurred (in the current time-reference frame),
  * no timestamps before `@LastSet` can be relied on even if the consumer understands epoch versioning and the timestamps are versioned because:
    *  the reason for the abrupt timestamp can't be determined by the participants, and
    * the timing for the abrupt time adjustment can't be determined by any participant to a resolution any higher than the sync timing.
* I don't think there's really any downside to setting `pm:ClockState/@ActivationState != "On"` but in the above context it isn't necessary. 
* We could reserve `Fail` for clock failures (e.g., hardware fault) and use `Stndby` for synchronization failures; maybe we don't need to separate these things though. 
* I think the strict part is in 1522 (providers must chose one of the options); 1569 is more about consumers understanding of what the fault state means; I assume providers can go into a fault state whenever they like. 
* Perhaps a `pm:AbstractDescriptor/@SafetyClassification="MedA"` (display only) might have a relaxed attitude to abrupt time jumps happy to just convey the information while a `MedC` (clinical function) provider might fault the whole device. 
* I better make 1569 refer to the provider as consumers don't have a activation state to set. 
* Is world time always a vital part of system function contribution? Perhaps a heart-lung machine would be more interested in continuing oxygenation than time-stamping exactly when a time-discontinuity occurred? 

If `mustUnderstand` isn't true, I don't think there is anything about how a consumer should interpret `ClockState/@ActivationState != "On"`. It might surprise them but SDPi conformance testing could validate they do something sensible. Although consumers that are on-board with SDPi should know what do with `mustUnderstand=true`. Are we worried about non SDPi compliant consumers? This just makes me think we should have `mustUnderstand=true`.


In conclusion, I'm not sure what to do next here.

Reply:
Ok, I think there is a lot of confusion going on - let my try to settle things:

- The requirement I listed above is needed to determine the time frame in which the clock is not trustworthy - it is not sufficient to formulate something along the lines of R1522, because it does only determine how to act at the time when there is a non-slewing adjustment. However, there was no restriction on the duration of keeping the activation state in StdBy
- Eventually, things need to be integratable in a way that consumers who do not support epoch versions can seemlessly rely on the StdBy to express incoherent timestamps - epoch versions is an extra feature consumers may use to adjust otherwise broken timestamps.

Now that I am getting depper into the topic, I think we can make the extension mustUnderstand=false, because we signify "something is being wrong" by using the StdBy flag and add epoch information to allow for consumers to calculate correct times based on epoch versions.

Reply:
This time business is much more complicated than it looks! I intended the latest R1522 to do all the things you mentioned:

---
When the <<vol1_spec_sdpi_p_actor_somds_provider>> detects an abrupt time adjustment of a system clock, used in making its System Function Contribution (<<acronym_sfc>>), the <<vol1_spec_sdpi_p_actor_somds_provider>> shall either:

1. initiate a new MDIB sequence by assigning a new MDIB sequence identifier, or
1. set `pm:ClockState/@ActivationState` to `StndBy` when any timestamp in a <<acronym_mdib>> version was not obtained from the time-reference frame of the active clock in the same version, or 
1. set `pm:ClockState/@LastSet` to the earliest time that is unambiguously in the current epoch and increment `sdpi:Epochs/@Version`.
---

To break it down:
* (1) is the sledgehammer, providing no information about the timeframe for invalid clock but may be a useful option for some providers. 
* (2) signals to the consumer, using `ClockState/@ActivationState`, that there are one or more timestamps from a different time-reference frame in the mdib. They could be anywhere (metrics, components or extensions etc) and there is no way to tell which timestamps are trustworthy and which are not. 
* (3) does two things:
  * it provides equivalent information to consumers as (2), though it does it using `sdpi:Epochs/@Version` changing and `pm:ClockState/@LastSet`. That is a consumer can compare every timestamp (whether metric, component or extension) to the value of `pm:ClockState/@LastSet`; any that are strictly less are untrustworthy. That's the case at the time of the adjustment or six months later. 
  * it lets consumers figure out which timestamps are untrustworthy (by comparing with `pm:ClockState/@LastSet`). 

> [!NOTE]
> the provider may generate timestamps from the current time-reference frame that are less than `pm:ClockState/@LastSet`. This can happen when a non-slewing adjustment moves the clock backwards in time and the MDIB previously contained timestamps from the "future". 


The downside of (3) not setting the `@ActivationState` is that the consumer has to:
1. understand `sdpi:Epochs/@Version`, and
2. check every timestamp for validity. 

One approach I can see providers taking with option 3 is leaving timestamps to update of their own accord. Notably, contexts may have untrustworthy timestamps until they are validated again, perhaps in a week or two when the provider is used on a new patient. It seems consumers might have a hard time figuring out what to do if `@ActivationState="StndBy"` for a week or two. However, if only setting `@ActivationState="StndBy"` is one of the approaches we want to have here, then I think it would make sense for the versioned one to also do that. 

So, I'll update the text to:

---
When the <<vol1_spec_sdpi_p_actor_somds_provider>> detects an abrupt time adjustment of a system clock, used in making its System Function Contribution (<<acronym_sfc>>), the <<vol1_spec_sdpi_p_actor_somds_provider>> shall either:

1. initiate a new MDIB sequence by assigning a new MDIB sequence identifier, or
1. set `pm:ClockState/@ActivationState` to `StndBy` when any timestamp in a <<acronym_mdib>> version was not obtained from the time-reference frame of the active clock in the same version, or 
1. set `pm:ClockState/@LastSet` to the earliest time that is unambiguously in the current epoch and increment `sdpi:Epochs/@Version` and set `pm:ClockState/@ActivationState` to `StndBy` while any timestamp in a <<acronym_mdib>> version is less than `pm:ClockState/@LastSet`.

---

I worded it slightly differently for greater clarity with the relationship to `pm:ClockState/@LastSet` but the behaviour of `pm:ClockState/@ActivationState` is identical for 2 & 3.